master/internal/scheduler/fair_share.go

```diff
@@ -159,7 +164,39 @@ func getTotalWeight(states []*groupState) float64 {
 	return total
 }
 
+func accountForPreoffers(preoffers int, offer int) (int, int) {
+	if preoffers > 0 {
+		if preoffers == offer {
+			preoffers = 0
+			offer = 0
+		}
+		if preoffers > offer {
+			preoffers -= offer
+			offer = 0
+		}
+		if preoffers < offer {
+			preoffers = 0
+			offer -= preoffers
+		}
+	}
+	return preoffers, offer
+}
+
 func allocateSlotOffers(states []*groupState, capacity int) {
+	// To prevent becoming oversubscribed, we first need to account for slots that were already
```

Comment:
I think we're good there: the notion of "preoffers" should ensure that once a non-preemptible task is scheduled, it continues to have enough offers as the highest priority. We already don't select a task to resolve a deadlock unless it failed to get enough offers. This might impact a non-preemptible task, but only before it gets scheduled, in which case it's fine for it to be treated the same as any other task. Do you agree?

Reply:
yeah, `smallestAllocatableTask.SlotsNeeded > state.offered` should always be false for those pre-offered task groups so they would not be preempted.

As I was trying to think of a corner case I found a type of case that is handled wrong. This is not caused by your change. The case is as follows:

You have two groups, the first group has an 8-slot task and an 1-slot task and the second group has three 2-slot tasks. You have two machines, one has 8 slots and the other one has 4 slots. Both groups would be offered 4 slots. The first group takes 4 slots but only start the 1-slot task while the second group takes 4 slots and starts two 2-slot tasks. There are 3 slots are idle while there is one two-slot task pending.

Check out the test case below:
```
func TestFairShareCornerCase(t *testing.T) {
	agents := []*mockAgent{
		{id: "agent1", slots: 8, label: ""},
		{id: "agent2", slots: 4, label: ""},
	}
	groups := []*mockGroup{
		{id: "group1"},
		{id: "group2"},
	}
	tasks := []*mockTask{
		{id: "task1", slotsNeeded: 8, group: groups[0]},
		{id: "task2", slotsNeeded: 1, group: groups[0]},
		{id: "task3", slotsNeeded: 2, group: groups[1]},
		{id: "task4", slotsNeeded: 2, group: groups[1]},
		{id: "task5", slotsNeeded: 2, group: groups[1]},
	}

	expectedToAllocate := []*mockTask{tasks[1], tasks[2], tasks[3], tasks[4]}
	expectedToRelease := []*mockTask{}

	system := actor.NewSystem(t.Name())
	taskList, groupMap, agentMap := setupClusterStates(t, system, tasks, groups, agents)
	toAllocate, toRelease := fairshareSchedule(taskList, groupMap, agentMap, BestFit)
	assertEqualToAllocate(t, toAllocate, expectedToAllocate)
	assertEqualToRelease(t, taskList, toRelease, expectedToRelease)
}
```

Reply:
Good point - thanks for reproducing that in a test. I've filed a follow-up issue to address that short-coming separately.